docs/repair_based_node_ops.md

```diff
@@ -0,0 +1,101 @@
+# Repair based node operations
+
+Here below is a simple introduction to the node operations Scylla support and
+some of the issues that streaming based node operations have.
+
+- Replace operation
+
+   It is used to replace a dead node. The token ring does not change. Replacing
+   node pulls data from only one of the replicas which might not be the latest
+   copy.
+
+- Rebuild operation
+
+   It is used to get all the data this node owns form other existing nodes. It
+   pulls data from only one of the replicas which might not be the latest copy.
+
+- Removenode operation
+
+   It is used to remove a dead node out of the cluster. Existing nodes pull
+   data from other existing nodes for the new ranges it owns. It pulls from one
+   of the replicas which might not be the latest copy.
+
+- Bootstrap operation
+
+   It is used to add a new node into the cluster. The token ring changes. It
+   does not suffer from the "latest replica” issue. New node pulls data
+   from existing nodes that are losing the token ranges.
+
+   It suffers from failed streaming. We split the ranges in 10 groups and we
+   stream one group at a time. Restream the group if failed, causing
+   unnecessary data transmission on wire.
+
+   Bootstrap is not resumable. With failure after 99.99% of data is streamed,
+   if we restart the node again, we need to stream all the data again even if
+   the node already has 99.99% of the data.
+
+- Decommission operation
+
+   It is used to remove a live node form the cluster. Token ring changes. It
+   does not suffer from the “latest replica” issue. The leaving node pushes data
+   to existing nodes.
+
+   It suffers from resumable issue like bootstrap operation.
+
+To solve all the issues above. We use repair based node operations. The idea
+behind repair based node operations is simple: use repair to synchronize data
+between replicas instead of using streaming.
+
+The benefits:
+
+- Latest copy is guaranteed
+
+- Resumable in nature
+
+- No extra data is streamed on wire
+  E.g., rebuild twice, will not stream the same data twice
+
+- Unified code path for all the node operations
+
+- Free repair operation during bootstrap, replace operation and so on.
+
+Select nodes to synchronize with:
+
+- Replace operation
+  Synchronize with all replica nodes from the local DC
+
+- Rebuild operation
+  Synchronize with all replica nodes from the local DC or from the DC specified by the user
+
+- Removenode operation
+  Synchronize with all replica nodes from local DC
+
+- Bootstrap operation
+  1) If local_dc_replica_nodes = RF, synchronize with 1 node that is losing the
+  range in the local DC
+
+  2) If 0 < local_dc_replica_nodes < RF, synchronize with
+  up to RF/2 + 1 nodes in local DC
+
+  3) If local_dc_replica_nodes = 0, reject the bootstrap operation
+
+- Decommission operation
+  1) Synchronize with one node that is gaining the range in local DC
+
```

Comment:
Again it may be down.

Reply:
The decommission will fail because the only neighbour is down. Like bootstrap, we reply on the fact the leaving node "transfers" the replica to the one gaining the range.

Reply:
But don't we have the option to either transfer to the node that is gaining the range, or to a quorum of other nodes?

Reply:
We do not have this option yet. Maybe a follower up patch.